pandas/core/reshape/melt.py

```diff
@@ -105,6 +107,7 @@ def melt(
     mdata = {}
     for col in id_vars:
         id_data = frame.pop(col)
+        assert isinstance(id_data, ABCSeries)  # help mypy
```

Comment:
This is 'fixing' the issue incorrectly.

The mypy error is `error: Incompatible types in assignment (expression has type "Union[DataFrame, Series]", variable has type "Series")` for L111 below

what's actually happening here is that ABCSeries resolves to Any and so effectively removing the variable type inferred for id_data (in the line above which is correctly inferred to be Series due to the typing added for DataFrame.pop)

the return type from concat is Union[DataFrame, Series]. 

to assign this to the same variable requires explicitly giving the variable that Union type (using a variable type annotation) or using a different variable name for subsequent assignments.

However, since the call to concat is with a sequence of Series and axis=0, we would expect the return type to be a Series not Union[DataFrame, Series]. To type concat to give the correct return type requires Literal which we cannot use at this time.

I _think_ the best solution is to cast the return value of the concat call for now. we can easily track our usage of cast.

I've just pushed the changes.

Reply:
Ah I see - massive thanks for taking the time to explain the issue with what I'd suggested, really appreciate it @simonjayhawkins